contrib/isa_l/compression/source/compressor/igzip_compressor_impl.cc

```diff
@@ -0,0 +1,91 @@
+#include "contrib/isa_l/compression/source/compressor/igzip_compressor_impl.h"
+
+#include <memory>
+
+#include "envoy/common/exception.h"
+
+#include "source/common/common/assert.h"
+
+#include "absl/container/fixed_array.h"
+#include "isa-l.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Compression {
+namespace Igzip {
+namespace Compressor {
+
+IgzipCompressorImpl::IgzipCompressorImpl() : IgzipCompressorImpl(4096) {}
+
+IgzipCompressorImpl::IgzipCompressorImpl(uint64_t chunk_size)
+    : Common::Base(chunk_size, [](isal_zstream* z) {
+        free(z->level_buf);
+        delete z;
+      }) {}
+
+void IgzipCompressorImpl::init(CompressionLevel comp_level, int64_t window_bits,
+                               int64_t memory_level) {
+  RELEASE_ASSERT(initialized_ == false, "");
+  isal_deflate_init(zstream_ptr_.get());
+  zstream_ptr_->avail_out = chunk_size_;
+  zstream_ptr_->next_out = chunk_char_ptr_.get();
+  zstream_ptr_->gzip_flag = IGZIP_GZIP;
+  zstream_ptr_->hist_bits = window_bits;
+
+  if (comp_level == CompressionLevel::Level1) {
+    zstream_ptr_->level = 1;
+  } else if (comp_level == CompressionLevel::Level2) {
+    zstream_ptr_->level = 2;
+  } else {
+    zstream_ptr_->level = 3;
+  }
+
+  uint64_t joint_index = (zstream_ptr_->level - 1) * 5 + memory_level - 1;
```

Comment:
I think we need some comment for this line? Or it just like a magic.

Reply:
Yeah, it's a magic to map compression level and memory level to the corresponding memory size.
I will add some comments to explain it next. Thanks!